### PR TITLE
distance along fixes for both node and edge searches

### PR DESCRIPTION
**Pull request by kevinkreiser**
fix two bugs one with node distance along and direction of the edge wrt the shape, which i had a todo about before and was warning my future self. and the corollary for the edge path

comments inline

### CODE REVIEW

File: src/loki/search.cc [hunk 1]
```diff
@@ -63,7 +63,7 @@ PathLocation CorrelateNode(const NodeInfo* node, const Location& location, const
     if(!filter(edge)) {
       GraphId id(tile->id());
       id.fields.id = node->edge_index() + (edge - start_edge);
-      correlated.CorrelateEdge(std::move(id), 0);
+      correlated.CorrelateEdge(std::move(id), static_cast<float>(!edge->forward()));
```
Comment (kevinkreiser) [inline src/loki/search.cc]: there used to be a comment here that said, should we check the edge direction, the answer was, and the test proves this, a resounding YES

File: src/loki/search.cc [hunk 2]
```diff
@@ -176,15 +176,24 @@ PathLocation EdgeSearch(const Location& location, GraphReader& reader, EdgeFilte
 
     //we haven't looked at this edge yet and its not junk
     if(!filter(edge) && visited.insert(edge->edgeinfo_offset()).second) {
+      //get the start and end node positions
+      auto end_node = GetEndNode(reader, edge);
+      auto start_node = GetBeginNode(reader, edge);
+
+      //is it basically right on the start of the edge
+      if(approximator.DistanceSquared(start_node->latlng()) < sq_node_snap) {
```
Comment (kevinkreiser) [inline src/loki/search.cc]: moved the node snapping up to the begging of this algorithm so that it can bail early

File: src/loki/search.cc [hunk 3]
```diff
@@ -241,6 +235,8 @@ PathLocation EdgeSearch(const Location& location, GraphReader& reader, EdgeFilte
         partial_length += closest_edge_info->shape()[i].Distance(closest_edge_info->shape()[i + 1]);
     partial_length += closest_edge_info->shape()[std::get<2>(closest_point)].Distance(std::get<0>(closest_point));
     float length_ratio = static_cast<float>(partial_length / static_cast<double>(closest_edge->length()));
+    if(!closest_edge->forward())
+      length_ratio = 1.f - length_ratio;
```
Comment (kevinkreiser) [inline src/loki/search.cc]: the corollary edge fix which inverts the distance ratio based on direction

File: test/search.cc [hunk 4]
```diff
@@ -71,17 +73,20 @@ boost::property_tree::ptree make_tile() {
...
   };
   auto add_edge = [&tile, &edge_index] (const std::pair<GraphId, PointLL>& u, const std::pair<GraphId, PointLL>& v,
-    const uint32_t name, const uint32_t opposing) {
+    const uint32_t name, const uint32_t opposing, const bool forward) {
 
     DirectedEdgeBuilder edge_builder;
     edge_builder.set_length(u.second.Distance(v.second) + .5);
     edge_builder.set_endnode(v.first);
     edge_builder.set_opp_index(opposing);
+    edge_builder.set_forward(forward);
```
Comment (kevinkreiser) [inline test/search.cc]: wasnt setting the direction in the graphtile building which is crucial for accurately testing this

File: src/loki/search.cc [hunk 5]
```diff
@@ -86,7 +86,7 @@ const NodeInfo* FindClosestNode(const Location& location, const GraphTile* tile,
   const auto end_node = start_node + tile->header()->nodecount();
   for(auto node = start_node; node < end_node; ++node) {
     //if this is closer then its better, unless nothing interesting leaves it..
-    float node_sqdist = approximator.DistanceSquared(location.latlng_);
+    float node_sqdist = approximator.DistanceSquared(node->latlng());
```
Comment (kevinkreiser) [inline src/loki/search.cc]: yeah this was a real bone head mistake

Comment (dnesbitt61) [inline src/loki/search.cc]: yea, easy to do though

**Comment by dnesbitt61** [comment]
:ship:

**Pull request merged**